### PR TITLE
[pnorm] optimize p_norm for special cases

### PR DESCRIPTION
**Pull request by LemonNoel**
### PR types

Performance optimization

### PR changes

OPs

### Describe

Optimize p_norm for two kinds of special cases:
(1) shape=[2, 1000, 1000], reduce axis=0
(2) shape=[1, 2000000, 1], reduce axis=1

The original version is paddlepaddle-gpu == 2.2.1. The ``Time`` denotes seconds per 1k steps.

- Forward Cases

| Tensor Shape | OP | Original Time| Current Time | Speedup |
| --- | --- | --- | --- | --- |
|[2, 1k, 1k] | paddle.norm(x, axis=0, p=3) | 4.0680373 | 0.0948312 | 45.2 | 
|[1, 2m, 1] | paddle.norm(x, axis=1, p=3) | 2.5773182 | 0.0548460 | 51.6 |
|[1m, 2] | paddle.norm(x, axis=1, p=3) | 4.0455515 | 0.0652089 | 57.9 |
|[1k, 1k] | paddle.norm(x, axis=0, p=3) | 0.0214734 | 0.0451496 | - |

- Backward Cases

| Shape | OP | Original Time | Current Time | Speedup | 
| --- | --- | --- | ---| --- | 
|[2, 1k, 1k] | paddle.norm(x, axis=0, p=3).sum().backward() | 4.6668179 | 0.3521516 | 13.3 |
|[1, 2m, 1] | paddle.norm(x, axis=1, p=3).sum().backward() | 4.8750155 | 0.2606518 |  18.8 |
|[1m, 2] | paddle.norm(x, axis=1, p=3).sum().backward() | 0.0756586 | 0.0661562 |  1.1 | 
|[1k, 1k] | paddle.norm(x, axis=0, p=3).sum().backward() | 0.0624404 | 0.0668254 | - |

### CODE REVIEW

**Comment by LemonNoel** [comment]
@Avin0323  你好，辛苦看下benchmark的CI。这个 PR 改动了 p_norm 的代码，优化了特殊 shape 的性能。从 CI 结果看 p_norm 的测试耗时都降低了。另外，还修改了 cmake 文件，辛苦 review 一下。

File: paddle/fluid/operators/p_norm_op.cu [hunk 1]
```diff
@@ -56,87 +59,94 @@ __device__ __forceinline__ double inline_pow(double base, double exponent) {
...
-      sum += static_cast<MT>(static_cast<double>(x_ij) != 0);
-    }
-    MT reduce_result = BlockReduce(temp_storage).Sum(sum);
-    if (threadIdx.x == 0) out_norm[i] = static_cast<T>(reduce_result);
+struct NonzeroFunctor {
+  HOSTDEVICE explicit inline NonzeroFunctor() {}
+  HOSTDEVICE explicit inline NonzeroFunctor(int n) {}
+  template <typename T>
+  HOSTDEVICE inline T operator()(const T& x) const {
+    return static_cast<T>(static_cast<double>(x) != 0);
```
Comment (ZHUI) [inline paddle/fluid/operators/p_norm_op.cu:76]: 为什么先 `static_cast<double>(x)` cast 为double？

File: paddle/fluid/operators/p_norm_op.cu [hunk 2]
```diff
@@ -146,101 +156,85 @@ class PnormCUDAKernel : public framework::OpKernel<T> {
...
     auto ndim = out_norm->dims();
     float porder = ctx.Attr<float>("porder");
     int axis = ctx.Attr<int>("axis");
     bool asvector = ctx.Attr<bool>("asvector");
     if (axis < 0) axis = xdim.size() + axis;
-    int pre, n, post;
-    GetDims(xdim, axis, &pre, &n, &post, asvector);
+    std::vector<int> reduce_axis = {axis};
 
     auto& dev_ctx = ctx.cuda_device_context();
```
Comment (ZHUI) [inline paddle/fluid/operators/p_norm_op.cu]: `dev_ctx` no usage in the function?

File: paddle/fluid/operators/p_norm_op.cu [hunk 3]
```diff
@@ -146,101 +156,85 @@ class PnormCUDAKernel : public framework::OpKernel<T> {
...
+                                               stream);
     } else if (porder == -INFINITY) {
-      NegInfNorm<T, block><<<grid, block, 0, dev_ctx.stream()>>>(x, pre, n,
-                                                                 post, norm);
+      TensorReduceFunctorImpl<T, T, AbsAndMin>(*in_x, out_norm, reduce_axis,
+                                               stream);
     } else {
-      Pnorm<T, block><<<grid, block, 0, dev_ctx.stream()>>>(x, pre, n, post,
-                                                            porder, norm);
+      framework::Tensor tmp_x;
```
Comment (ZHUI) [inline paddle/fluid/operators/p_norm_op.cu:179]: 记一下todo，这里的 `tmp_x` 需要尽早去掉。运行时显存占用提升很多。

File: paddle/fluid/operators/p_norm_op.cu [hunk 4]
```diff
@@ -146,101 +156,85 @@ class PnormCUDAKernel : public framework::OpKernel<T> {
...
+  template <typename DeviceContext, typename X, typename Y, typename DX,
+            typename DY, typename Dim>
+  void operator()(const DeviceContext& place, X* x, Y* y, DX* dx, DY* dy,
+                  const Dim& dim, int size) {
+    auto equals = ((*x).abs() == y->broadcast(dim));
+    auto ones = dx->constant(static_cast<T>(1.));
+    auto negs = dx->constant(static_cast<T>(-1.));
+    auto zeros = dx->constant(static_cast<T>(0.));
+    auto positives = (*x) > zeros;
+    dx->device(place) = dy->broadcast(dim) * equals.select(ones, zeros) *
```
Comment (ZHUI) [inline paddle/fluid/operators/p_norm_op.cu:217]: 这里反向，都是走的eigen？

File: paddle/fluid/operators/p_norm_op.cu [hunk 5]
```diff
@@ -260,32 +254,38 @@ class PnormGradCUDAKernel : public framework::OpKernel<T> {
     float porder = ctx.Attr<float>("porder");
     T eps = static_cast<T>(ctx.Attr<float>("epsilon"));
     int axis = ctx.Attr<int>("axis");
+    bool reduce_all = ((axis < 0) || (in_norm->numel() == 1));
```
Comment (ZHUI) [inline paddle/fluid/operators/p_norm_op.cu:251]: `axis < 0` 是对应 `reduce_all` 吗？

File: paddle/fluid/operators/p_norm_op.cu [hunk 6]
```diff
@@ -260,32 +254,38 @@ class PnormGradCUDAKernel : public framework::OpKernel<T> {
...
     bool asvector = ctx.Attr<bool>("asvector");
     if (axis < 0) axis = xdim.size() + axis;
-    int pre, n, post;
-    GetDims(xdim, axis, &pre, &n, &post, asvector);
+    const std::vector<int> dims = {axis};
 
     auto& dev_ctx = ctx.cuda_device_context();
```
Comment (ZHUI) [inline paddle/fluid/operators/p_norm_op.cu]: `dev_ctx` 是否还有使用？

**Review by ZHUI — DISMISSED** [review]

File: paddle/fluid/operators/p_norm_op.cu (hunk 5 above)
Comment (LemonNoel) [reply paddle/fluid/operators/p_norm_op.cu:251]: 是的

File: paddle/fluid/operators/p_norm_op.cu (hunk 1 above)
Comment (LemonNoel) [reply paddle/fluid/operators/p_norm_op.cu:76]: 这里保留了原始实现

File: paddle/fluid/operators/p_norm_op.cu (hunk 2 above)
Comment (LemonNoel) [reply paddle/fluid/operators/p_norm_op.cu]: 是的，晚点提PR删掉

File: paddle/fluid/operators/p_norm_op.cu (hunk 3 above)
Comment (LemonNoel) [reply paddle/fluid/operators/p_norm_op.cu:179]: TODO 好的，已提交卡片

File: paddle/fluid/operators/p_norm_op.cu (hunk 6 above)
Comment (LemonNoel) [reply paddle/fluid/operators/p_norm_op.cu]: 已删除

File: paddle/fluid/operators/p_norm_op.cu (hunk 4 above)
Comment (LemonNoel) [reply paddle/fluid/operators/p_norm_op.cu:217]: 计算是eigen tensor

**Review by Avin0323 — APPROVED** [review]
LGTM for PR-CI-OP-benchmark and changes of unity_build_rule.cmake

**Review by ZHUI — APPROVED** [review]
LGTM

**Pull request merged**